File: py_test/second/ping_to_ping_associationRoee_test.py
import numpy as np
import csv
import cov_plot_calcRoee as cpc
import std_calcRoee as scr
import ping_to_ping_associationRoee as p2par  # Adjust import based on the actual function location

def load_csv_to_array(filename):
    with open(filename, 'r') as f:
        reader = csv.reader(f)
        data = [row for row in reader]
    return np.array(data, dtype=float), len(data)

def load_csv_to_2d_array(filename):
    with open(filename, 'r') as f:
        reader = csv.reader(f)
        data = []
        for row in reader:
            # Strip any trailing empty elements caused by extra commas
            row = [x for x in row if x.strip()]

            if any(row):  # Check if the row contains any non-empty elements
                data.append([float(x) if x else 0.0 for x in row])  # Handle empty values

    return np.array(data, dtype=float), len(data)

def load_csv_to_3d_array2(filename):
    data = []
    current_slice = []
    
    with open(filename, 'r') as f:
        reader = csv.reader(f)
        for row in reader:
            if row:  # If the row is not empty
                current_slice.append([float(x) for x in row])
            else:  # Empty row indicates a new slice
                if current_slice:
                    data.append(current_slice)
                    current_slice = []
                    
        if current_slice:  # Add the last slice if the file does not end with an empty row
            data.append(current_slice)

    return np.array(data), len(data)

def load_csv_to_3d_array(filename):
    data = []
    current_slice = []

    with open(filename, 'r') as f:
        reader = csv.reader(f)
        for row in reader:
            # Strip any trailing empty elements caused by extra commas at the end of the line
            row = [x for x in row if x.strip()]

            if row:  # If the row contains any valid elements
                current_slice.append([float(x) for x in row])
            else:
                # Empty row indicates a new 4x4 matrix slice
                if current_slice:
                    data.append(current_slice)
                    current_slice = []  # Reset for the next slice

        # Append the last slice if it doesn't end with an empty row
        if current_slice:
            data.append(current_slice)

    # Convert the data into a numpy array (3D) and return the shape
    reshaped_data = np.array(data)
    return reshaped_data, reshaped_data.shape

def load_csv_to_1d_array(filename):
    with open(filename, 'r') as f:
        reader = csv.reader(f)
        data = []
        for row in reader:
            # Filter out empty strings and convert valid values to float
            filtered_row = [float(x) if x.strip() else 0.0 for x in row]  # Handle empty values as 0.0
            data.append(filtered_row)
    return np.array(data, dtype=float), len(data)

def test_ping_to_ping_associationRoee():
  
    # Load inputs from CSV files
    
    # Before path
    before_path = "C:\\Users\\ipewz\\Desktop\\roee_proj\\forTest\\ping_to_ping_correlationRoee_befor\\"
    
    MergedRng, _ = load_csv_to_2d_array(before_path + "MergedRng.csv")
    MergedTeta, _ = load_csv_to_2d_array(before_path + "MergedTeta.csv")
    MergedYc, _ = load_csv_to_2d_array(before_path + "MergedYc.csv")
    


    # After path
    associationRoee_after_path = "C:\\Users\\ipewz\\Desktop\\roee_proj\\forTest\\ping_to_ping_associationRoee_after\\"
    
    TracksDataBinMat_after  , _ = load_csv_to_2d_array(associationRoee_after_path + "TracksDataBinMat.csv")
    TracksDataMat_after     , _ = load_csv_to_2d_array(associationRoee_after_path + "TracksDataMat.csv")
    TracksMat_after         , _ = load_csv_to_2d_array(associationRoee_after_path + "TracksMat.csv")
    TracksMissMat_after     , _ = load_csv_to_3d_array(associationRoee_after_path + "TracksMissMat.csv")
    TracksP_after           , _ = load_csv_to_3d_array(associationRoee_after_path + "TracksP.csv")
    TracksVecMat_after      , _ = load_csv_to_2d_array(associationRoee_after_path + "TracksVecMat.csv")
    TracksX_after           , _ = load_csv_to_2d_array(associationRoee_after_path + "TracksX.csv")
    
    
    

    # ping_to_ping_correlationRoee_after - is - ping_to_ping_associationRoee_before
    after_path = "C:\\Users\\ipewz\\Desktop\\roee_proj\\forTest\\ping_to_ping_correlationRoee_after\\"

    TracksDataBinMat  , _ = load_csv_to_2d_array(after_path + "TracksDataBinMat.csv")
    TracksDataMat     , _ = load_csv_to_2d_array(after_path + "TracksDataMat.csv")
    TracksMat         , _ = load_csv_to_2d_array(after_path + "TracksMat.csv")
    TracksMissMat     , _ = load_csv_to_3d_array(after_path + "TracksMissMat.csv")
    TracksP           , _ = load_csv_to_3d_array(after_path + "TracksP.csv")
    TracksVecMat      , _ = load_csv_to_2d_array(after_path + "TracksVecMat.csv")
    TracksX           , _ = load_csv_to_2d_array(after_path + "TracksX.csv")
    #uncorr_plots_list , _ = load_csv_to_1d_array(after_path + "uncorr_plots_list.csv")


    NumDetect = 13  # Set this based on your test data
    ping_ind = 1
    Tping = 1
    sigmaTeta = 3  # Example value for sigmaTeta
    ping_number = 2
    cov_fact =5 
    Win_dlt = 10
    xmax = 10
    CurrentTargetInd = 10
    CurrentTargetInd_after = 12

    # Call the function
   
   
    (
    TracksMat, 
    TracksVecMat, 
    TracksP, 
    TracksX, 
    TracksMissMat, 
    TracksDataBinMat, 
    TracksDataMat, 
    CurrentTargetInd
    )  = p2par.ping_to_ping_associationRoee(
        TracksMat, 
        TracksVecMat, 
        TracksP, 
        TracksX, 
        TracksMissMat, 
        TracksDataBinMat, 
        TracksDataMat, 
        MergedRng[0], 
        MergedTeta[0], 
        MergedYc, 
        NumDetect, 
        ping_number, 
        sigmaTeta, 
        cov_fact, 
        Win_dlt, 
        xmax, 
        Tping, 
        CurrentTargetInd
        )

    
         # Assertions
    '''
    assert np.array_equal(TracksMat, TracksMat_after), "TracksMat_out does not match expected values"
    print("tests 1 passed")
    assert np.array_equal(TracksVecMat, TracksVecMat_after), "TracksVecMat_out does not match expected values"
    print("tests 2 passed")
    assert np.array_equal(TracksP,TracksP_after ), "TracksP_out does not match expected values"
    print("tests 3 passed")
    assert np.array_equal(TracksX, TracksX_after), "TracksX_out does not match expected values"
    print("tests 4 passed")
    assert np.array_equal(TracksMissMat, TracksMissMat_after), "TracksMissMat_out does not match expected values"
    print("tests 5 passed")
    assert np.array_equal(TracksDataBinMat, TracksDataBinMat_after), "TracksDataBinMat_out does not match expected values"
    print("tests 6 passed")
    assert np.array_equal(TracksDataMat, TracksDataMat_after), "TracksDataMat_out does not match expected values"
    print("tests 7 passed")
    assert np.array_equal(CurrentTargetInd, CurrentTargetInd_after), "uncorr_plots_list_out does not match expected values"
    #print("tests 8 passed")
    print("tests 8 is ok but it show err because of matlab start from 1 and py from 0")
    print("All tests passed")
    '''


    try:
        assert np.array_equal(TracksMat, TracksMat_after), "TracksMat_out does not match expected values"
        print("Test 1 passed")
    except AssertionError as e:
        print(f"Test 1 failed: {e}")

    try:
        assert np.array_equal(TracksVecMat, TracksVecMat_after), "TracksVecMat_out does not match expected values"
        print("Test 2 passed")
    except AssertionError as e:
        print(f"Test 2 failed: {e}")

    try:
        assert np.array_equal(TracksP, TracksP_after), "TracksP_out does not match expected values"
        print("Test 3 passed")
    except AssertionError as e:
        print(f"Test 3 failed: {e}")

    try:
        assert np.array_equal(TracksX, TracksX_after), "TracksX_out does not match expected values"
        print("Test 4 passed")
    except AssertionError as e:
        print(f"Test 4 failed: {e}")

    try:
        assert np.array_equal(TracksMissMat, TracksMissMat_after), "TracksMissMat_out does not match expected values"
        print("Test 5 passed")
    except AssertionError as e:
        print(f"Test 5 failed: {e}")

    try:
        assert np.array_equal(TracksDataBinMat, TracksDataBinMat_after), "TracksDataBinMat_out does not match expected values"
        print("Test 6 passed")
    except AssertionError as e:
        print(f"Test 6 failed: {e}")

    try:
        assert np.array_equal(TracksDataMat, TracksDataMat_after), "TracksDataMat_out does not match expected values"
        print("Test 7 passed")
    except AssertionError as e:
        print(f"Test 7 failed: {e}")
    
    try:
        assert np.array_equal(CurrentTargetInd, CurrentTargetInd_after), "CurrentTargetInd does not match expected values"
        print("Test 8 passed")
    except AssertionError as e:
        print(f"Test 8 failed: {e}")


if __name__ == "__main__":
    test_ping_to_ping_associationRoee()
    
    

File: c/secund/ping_to_ping_associationRoee_test.c
#include "ping_to_ping_associationRoee.h"

#include <assert.h>
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Function to load a 2D array from a CSV file
void load_csv_to_2d_array(const char* filename, double* data, int rows, int cols) {
    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        perror("Unable to open file");
        exit(EXIT_FAILURE);
    }

    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            fscanf(file, "%lf,", &data[i * cols + j]);
        }
    }

    fclose(file);
}

// Function to load a 3D array from a CSV file
void load_csv_to_3d_array(const char* filename, double* data, int slices, int rows, int cols) {
    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        perror("Unable to open file");
        exit(EXIT_FAILURE);
    }

    for (int k = 0; k < slices; k++) {
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                fscanf(file, "%lf,", &data[k * rows * cols + i * cols + j]);
            }
        }
    }

    fclose(file);
}

// Function to load a 1D array from a CSV file
void load_csv_to_1d_array(const char* filename, double* data, int size) {
    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        perror("Unable to open file");
        exit(EXIT_FAILURE);
    }

    for (int i = 0; i < size; i++) {
        fscanf(file, "%lf,", &data[i]);
    }

    fclose(file);
}

// Function to compare two arrays and assert their equality
void assert_arrays_equal(double* arr1, double* arr2, int size, const char* msg, double tol) {
    for (int i = 0; i < size; i++) {
        if (fabs(arr1[i] - arr2[i]) > tol) {
            printf("%s: Arrays differ at index %d (%.6f != %.6f)\n", msg, i, arr1[i], arr2[i]);
            exit(EXIT_FAILURE);
        }
    }
    printf("%s: Test passed\n", msg);
}

// Function to check if two arrays are nearly equal (for float comparison)
bool arrays_are_close(double* arr1, double* arr2, int size, double tol) {
    for (int i = 0; i < size; i++) {
        if (fabs(arr1[i] - arr2[i]) > tol) {
            return false;
        }
    }
    return true;
}

// Function to check if two matrices are nearly equal (for float comparison)
bool matrices_are_close(double matrix1[4][4], double matrix2[4][4], int rows, int cols, double tol) {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (fabs(matrix1[i][j] - matrix2[i][j]) > tol) {
                return false;
            }
        }
    }
    return true;
}

void test_ping_to_ping_associationRoee() {
    int NumDetect = 13;
    int ping_ind = 1;
    int Tping = 1;
    int sigmaTeta = 3;
    int ping_number = 2;
    int cov_fact = 5;
    int Win_dlt = 10;
    int xmax = 10;
    int CurrentTargetInd = 10;
    int CurrentTargetInd_after = 12;

    // Load input data
    double* TracksMat = (double*)malloc(3 * 360 * sizeof(double));
    double* TracksVecMat = (double*)malloc(4 * 360 * sizeof(double));
    double* TracksP = (double*)malloc(360 * 4 * 4 * sizeof(double));
    double* TracksX = (double*)malloc(4 * 360 * sizeof(double));
    double* TracksMissMat = (double*)malloc(5 * 360 * 15 * sizeof(double));
    double* TracksDataBinMat = (double*)malloc(360 * 360 * sizeof(double));
    double* TracksDataMat = (double*)malloc(360 * 360 * sizeof(double));
    double* MergedRng = (double*)malloc(NumDetect * sizeof(double));
    double* MergedTeta = (double*)malloc(NumDetect * sizeof(double));
    double* MergedYc = (double*)malloc(NumDetect * 2 * sizeof(double));

    load_csv_to_1d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_correlationRoee_befor/MergedRng.csv", MergedRng, NumDetect);
    load_csv_to_1d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_correlationRoee_befor/MergedTeta.csv", MergedTeta, NumDetect);
    load_csv_to_2d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_correlationRoee_befor/MergedYc.csv", MergedYc, NumDetect, 2);
    load_csv_to_2d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_correlationRoee_after/TracksDataBinMat.csv", TracksDataBinMat, 360, 360);
    load_csv_to_2d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_correlationRoee_after/TracksDataMat.csv", TracksDataMat, 360, 360);
    load_csv_to_2d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_correlationRoee_after/TracksX.csv", TracksX, 4, 360);
    load_csv_to_2d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_correlationRoee_after/TracksVecMat.csv", TracksVecMat, 4, 360);
    load_csv_to_2d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_correlationRoee_after/TracksMat.csv", TracksMat, 3, 360);
    load_csv_to_3d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_correlationRoee_after/TracksMissMat.csv", TracksMissMat, 5, 360, 15);
    load_csv_to_3d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_correlationRoee_after/TracksP.csv", TracksP, 360, 4, 4);

    // Load expected output data
    double* TracksMat_after = (double*)malloc(3 * 360 * sizeof(double));
    double* TracksVecMat_after = (double*)malloc(4 * 360 * sizeof(double));
    double* TracksP_after = (double*)malloc(360 * 4 * 4 * sizeof(double));
    double* TracksX_after = (double*)malloc(4 * 360 * sizeof(double));
    double* TracksMissMat_after = (double*)malloc(5 * 360 * 15 * sizeof(double));
    double* TracksDataBinMat_after = (double*)malloc(360 * 360 * sizeof(double));
    double* TracksDataMat_after = (double*)malloc(360 * 360 * sizeof(double));

    load_csv_to_2d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_associationRoee_after/TracksDataBinMat.csv", TracksDataBinMat_after, 360, 360);
    load_csv_to_2d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_associationRoee_after/TracksDataMat.csv", TracksDataMat_after, 360, 360);
    load_csv_to_3d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_associationRoee_after/TracksMissMat.csv", TracksMissMat_after, 5, 360, 15);
    load_csv_to_3d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_associationRoee_after/TracksP.csv", TracksP_after, 360, 4, 4);
    load_csv_to_2d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_associationRoee_after/TracksVecMat.csv", TracksVecMat_after, 4, 360);
    load_csv_to_2d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_associationRoee_after/TracksX.csv", TracksX_after, 4, 360);
    load_csv_to_2d_array("C:/Users/ipewz/Desktop/roee_proj/forTest/ping_to_ping_associationRoee_after/TracksMat.csv", TracksMat_after, 3, 360);

    // Call the function
    ping_to_ping_associationRoee(TracksMat, TracksVecMat, TracksP, TracksX, TracksMissMat, TracksDataBinMat, TracksDataMat, MergedRng, MergedTeta, MergedYc, NumDetect, ping_ind, sigmaTeta, cov_fact, Win_dlt, xmax, Tping, CurrentTargetInd);
/*
    printf("\nTracksX:       ");
    for (size_t i = 0; i < 360; i++) {
        if (*(TracksX + 360 + i) < 100) {
            printf("%f ", *(TracksX + 360 + i));
        }
    }

    printf("\nTracksX_after  ");
    for (size_t i = 0; i < 360; i++) {
        if (*(TracksX_after + 360 + i) < 100) {
            printf("%f ", *(TracksX_after + 360 + i));
        }
    }

    printf("\nTracksVecMat:       ");
    for (size_t j = 0; j < 4; j++) {
        printf("\n");
        for (size_t i = 0; i < 360; i++) {
            if (*(TracksVecMat + j * 360 + i) < 900) {
                printf("%f ", *(TracksVecMat + j * 360 + i));
            }
        }
    }

    printf("\nTracksVecMat_after  ");
    for (size_t j = 0; j < 4; j++) {
        printf("\n");
        for (size_t i = 0; i < 360; i++) {
            if (*(TracksVecMat_after + j * 360 + i) < 900) {
                printf("%f ", *(TracksVecMat_after + j * 360 + i));
            }
        }
    }
*/
/*
    printf("\n TracksP:       ");
    for (size_t j = 0; j < 360; j++) {
        printf("\n");
        for (size_t i = 0; i < 16; i++) {
         //   if (*(TracksDataMat + j * 360 + i) < 900) {
                printf("%f ", *(TracksP + j * 16 + i));
                printf("%f ", *(TracksP_after + j * 16 + i));
          //  }
        }
    }
*/
 /*printf("\n TracksMissMat:       ");
    for (size_t j = 0; j < 5; j++) {
       // printf("\n");
        for (size_t i = 0; i < 360; i++) {
         printf("\n");
        for (size_t k = 0; k < 15; k++) {
         //   if (*(TracksDataMat + j * 360 + i) < 900) {
                printf("%f ", *(TracksMissMat + j* 360 * 15+ i * 360+k));
                printf("%f ", *(TracksMissMat_after + j* 360 * 15+ i * 360+k));
          //  }
        }}
    }
*/
    // Tolerance for float comparison
    double tolerance = 0.1;
    printf("\n-------------------------------\n");

    if (!arrays_are_close(TracksDataMat, TracksDataMat_after, 360 * 360, tolerance)) {
        printf("\nTest failed: TracksDataMat does not match expected values\n");
    } else {
        printf("\nTest passed: TracksDataMat matches expected values\n");
    }
    if (!arrays_are_close(TracksDataBinMat, TracksDataBinMat_after, 360 * 360, tolerance)) {
        printf("\nTest failed: TracksDataBinMat does not match expected values\n");
    } else {
        printf("\nTest passed: TracksDataBinMat matches expected values\n");
    }

    if (!arrays_are_close(TracksVecMat, TracksVecMat_after, 360 * 4, tolerance)) {
        printf("\nTest failed: TracksVecMat does not match expected values\n");
    } else {
        printf("\nTest passed: TracksVecMat matches expected values\n");
    }
    if (!arrays_are_close(TracksP, TracksP_after, 360 * 4*4, tolerance)) {
        printf("\nTest failed: TracksP does not match expected values\n");
    } else {
        printf("\nTest passed: TracksP matches expected values\n");
    }
    if (!arrays_are_close(TracksMissMat, TracksMissMat_after,5*360*15, tolerance)) {
        printf("\nTest failed: TracksMissMat does not match expected values\n");
    } else {
        printf("\nTest passed: TracksMissMat matches expected values\n");
    }
    if (!arrays_are_close(TracksMat, TracksMat_after, 360 * 3, tolerance)) {
        printf("\nTest failed: TracksMat does not match expected values\n");
    } else {
        printf("\nTest passed: TracksMat matches expected values\n");
    }
    if (!arrays_are_close(TracksX, TracksX_after, 360 * 4, tolerance)) {
        printf("\nTest failed: TracksX does not match expected values\n");
    } else {
        printf("\nTest passed: TracksX matches expected values\n");
    }
    printf("\n-------------------------------\n");

    // Compare the outputs with the expected data
    assert_arrays_equal(TracksDataMat, TracksDataMat_after, 360 * 360, "\nTracksDataMat",tolerance);
    printf("\ntests 1 passed!\n");
    assert_arrays_equal(TracksDataBinMat, TracksDataBinMat_after, 360 * 360, "\nTracksDataBinMat",tolerance);
    printf("\ntests 2 passed!\n");
    assert_arrays_equal(TracksMat, TracksMat_after, 360 * 3, "\nTracksMat",tolerance);
    printf("tests 3 passed!\n");
    assert_arrays_equal(TracksMissMat, TracksMissMat_after, 5 * 360 * 15, "\nTracksMissMat",tolerance);
    printf("tests 4 passed!\n");
    assert_arrays_equal(TracksP, TracksP_after, 360 * 4 * 4, "\nTracksP",tolerance);
    printf("tests 5 passed!\n");
    assert_arrays_equal(TracksVecMat, TracksVecMat_after, 360 * 4, "\nTracksVecMat",tolerance);
    printf("tests 6 passed!\n");
    assert_arrays_equal(TracksX, TracksX_after, 360 * 4, "\nTracksX",tolerance);

    printf("\nAll tests passed!\n");
}

int main() {
    test_ping_to_ping_associationRoee();
    return 0;
}

/*

 gcc ping_to_ping_associationRoee_test.c -o ping_to_ping_associationRoee_test.exe -lm
./ping_to_ping_associationRoee_test.exe

*/
